### ISSUE TITLE
"Accept" header should be "application/json" when getting token

### ISSUE THREAD

**Issue by giddie** [ueberauth/oauth2#165]
**Version:** 2.0.0

According to [the OAuth spec (Section 4.1.4)](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.4), the "Accept" header should be "application/json" when requesting a token. However, the default is currently "application/x-www-form-urlencoded" when calling `Client.get_token()`.

The 3rd-party OAuth server I'm working with rejects the request with **406 Not Acceptable**. If I add `{"accept", "application/json"}` to the `headers` param when calling `Client.get_token()`, then the request succeeds. The request also succeeds if the "Accept" header is empty or missing.

**Comment by yordis** [comment]
@giddie are you able to contribute a fix for it?

**Comment by giddie** [comment]
Sure thing :+1:

**Comment by yordis** [comment]
@giddie first, somehow I completely miss that you created a PR a long time ago, my apologies.

Re-reading the [the OAuth spec (Section 4.1.4)](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.4) I can't find where it says that it is required to set such header. Do you mind linking to the precise paragraph and maybe taking some screenshots? 

I really don't want to miss and I re-read it multiple times.

My apologies for the extra work, really appreciate your support.

**Comment by yordis** [comment]
My concern is that, kind of related to https://github.com/ueberauth/oauth2/issues/164 you can have a bit more control over the content type and accept, so I am not sure if "assuming" that everyone must be `application/json` is a good call or not.

Again, my apologies, I need to take the time to read the entire spec, so more than welcome to share some bits here about it.

**Comment by giddie** [comment]
As far as I can tell, when a client sends a request for a token to a server, a missing "Accept" header would be just fine according to the spec. The examples provided in the spec don't have an Accept header. However:

> The parameters are included in the entity-body of the HTTP response using the "application/json" media type as defined by [[RFC4627](https://datatracker.ietf.org/doc/html/rfc4627)].
  --- 5.1 Successful Response

So an Accept header of "application/x-www-form-urlencoded", sent by the client, is _definitely_ wrong, because by honouring the requested media type, the server would be breaking the spec. I think the simplest solution is just to ensure that the "Accept" header is "application/json" for this request.

**Comment by yordis** [comment]
> However, when the server responds with the token:

Right, which it sounds fine to me, that is the server telling the client the content-type of the response.

> So an Accept header of "application/x-www-form-urlencoded", sent by the client, is definitely wrong,

Where is that happening? 

I see we send the `content-type` which is different: https://github.com/ueberauth/oauth2/blob/0e05de4441706631017ee05e93fe01ec75e0dea2/lib/oauth2/client.ex#L483

So I am a bit confused about it.

> because by honouring the requested content type, the server would be breaking the spec.

Totally get that, and it sounds like a good optimization, but it seems that the server should `not` honor or do whatever they want with such information, `unless` it is not defined like that in the spec.

**Comment by giddie** [comment]
> > So an Accept header of "application/x-www-form-urlencoded", sent by the client, is definitely wrong,
> 
> Where is that happening?

Right here, I believe:

https://github.com/ueberauth/oauth2/blob/00640c3809faba5955029a2821c8ab60344044f4/lib/oauth2/request.ex#L131

**Comment by yordis** [comment]
Following that code, it is adding `content_type` if no `accept` header was specified. The `content_type` will be the `Content-Type` if it was specified.

I think it is OK to do such a change since the spec says it must return `application/json` content type anyway

**Comment by giddie** [comment]
> Following that code, it is adding `content_type` if no `accept` header was specified. The `content_type` will be the `Content-Type` if it was specified.

Not quite: I believe it's setting the `Accept` header to match the `Content-Type` if it's not already defined. In this case, the content type when requesting a token will be `application/x-www-form-urlencoded`, whereas the response will be `application/json`. So this fallback behaviour must be bypassed in one way or another for the token request.
